Doc: Adding documentation for multi-arch operators (#5178)
operator-framework/operator-sdk: website/content/en/docs/building-operators/golang/advanced-topics.md

When the operator is not running in a cluster, the Manager will return an error on starting since it can't detect the operator's namespace in order to create the configmap for leader election. You can override this namespace by setting the Manager's `LeaderElectionNamespace` option.

> ### Multiple architectures

Reviewer: +1 for it is not in the Golang advanced topics. It seems valid for all languages. It would better fit as a section such as Scorecard (ihmo)

Author: I am happy to do that if there is consensus for it. @asmacdo @joelanford what is your take on this?

Operators may run on various architectures: x86_64, aarch64, ppc64le, s390x, etc.

> 

Reviewer: What about add when it is useful?

Author: Added some text for it

#### Manifest lists

The most straightforward way of building operators and operands supporting multiple architectures is to leverage manifest lists, specified by [Image Manifest V2, Schema 2](https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list) or [OCI Image Index](https://github.com/opencontainers/image-spec/blob/main/image-index.md). A manifest list points to specific image manifests for one or more platforms.

> For convenience tools like [buildah](https://github.com/containers/buildah/blob/main/docs/buildah-bud.md#building-an-multi-architecture-image-using-a---manifest-option-requires-emulation-software) allow to cross-build and manifest multi-arch containers on one host. For instance:

Reviewer: What about adding the links at the bottom of the page? 
See that is the common standard in the docs to make it easier to keep them maintained.

Author: I am not convinced it is useful but I made the change to keep it consistent

> 

Reviewer: ```suggestion
**By using buildah tool** 
```

```shell
for a in amd64 arm64 ppc64le s390x; do \
  buildah bud --manifest registry/username/repo:v1 --arch $a; \
done
```

This creates the manifest list, builds each image and adds them to the manifest list.

The result can then be pushed to the desired registry.

```shell
buildah push registry/username/repo:v1
```

> 

Reviewer: ```suggestion
**By using Docker tool** 
```

Docker with [buildx](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) provides similar capabilities.

```shell
docker buildx build --push --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x --tag registry/username/repo:v1 .
```

See [docker documentation](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) for additional options.
 

> **Caveats**: the Dockerfile generated by the SDK for the operator explicitely references GOARCH=amd64 for go build. This can be amended to GOARCH=$TARGETARCH. Docker will automatically set the environment variable to the value specified by --platform. With buildah --build-arg will need to be used for the purpose.

Reviewer: ```suggestion
**NOTE**: the Dockerfile generated by the SDK for the operator explicitly references `GOARCH=amd64` for go build. This can be amended to `GOARCH=$TARGETARCH`. Docker will automatically set the environment variable to the value specified by --platform. With `buildah --build-arg` will need to be used for the purpose.
```

Author: IMHO this is more a caveat or warning than a note.

> **Caveats**: the Dockerfile generated by the SDK for the operator explicitely references GOARCH=amd64 for go build. This can be amended to GOARCH=$TARGETPLATFORM. Docker will automatically set the environment variable to the value specified by --platform. With buildah --build-arg will need to be used for the purpose.

Reviewer 2: Should this be:
```suggestion
**Caveats**: the Dockerfile generated by the SDK for the operator explicitely references GOARCH=amd64 for go build. This can be amended to GOARCH=$TARGETARCH. Docker will automatically set the environment variable to the value specified by --platform. With buildah --build-arg will need to be used for the purpose.
```

I'm pretty sure TARGETPLATFORM would expand to `linux/amd64` (or whatever the full os/arch combo was)

Author: Good catch

> All image references in the ClusterServiceVersion should be manifest lists containing the pointers to the image manifests for the supported architectures.

Reviewer: ```suggestion
All images references in the ClusterServiceVersion should be manifest lists containing the pointers to the image manifests for the supported architectures.
```

Reviewer: (ihmo)Here needs to be in sub-section such as `Using mult-arch in the bundle` or `Using multi-arch with the OLM integration`

Author: "all image references" is correct

Author: Moved the line to the CSV part

> All image references in the ClusterServiceVersion should be manifest lists containing the pointers to the image manifests for the supported architectures.

Reviewer: WDYT about the sections be more related to the tasks and possible goals? E.g Integrating with OLM, Generating images with other archs and etc...

Author: Moved bundle and CSV into an OLM subsection.

> **Caveats**: When mirroring registries for disconnected installations all the images referenced by a manifest list need to be copied, including images for architectures that may not be used in the environment.

Reviewer: ```suggestion
**Notes**: When mirroring registries for disconnected installations all the images referenced by a manifest list need to be copied, including images for architectures that may not be used in the environment.
```

Reviewer: Is that valid for upstream or only OCP downstream?

Reviewer: We might also link what means disconnect and etc..

Author: Added that environments without internet connection are meant by that

#### Bundle image

> Bundle images are not architecture-specific. They contain only plaintext kubernetes manifests and operator metadata.

Reviewer: iihmo we also need to link bundle here with the SDK doc that describes what is a bundle and how to integrate the project with OLM. 
```suggestion
Bundle images are not architecture-specific. They contain only plaintext Kubernetes manifests and operator metadata.
```

Author: added link to bundle documentation

#### ClusterServiceVersion labels

> 

Reviewer: ```suggestion
```
nit: addtional space

Labels for OS and architectures can be set in the CSV. Please refer to the [Operator Lifecycle Management Documentation](https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/#multiple-architectures) for details.

> 

Reviewer: we need to clarifies that if the user is building projects to integrate and distribute via OLM then ...
Not all SDK users know about OLM or are using it.

When the operator is not running in a cluster, the Manager will return an error on starting since it can't detect the operator's namespace in order to create the configmap for leader election. You can override this namespace by setting the Manager's `LeaderElectionNamespace` option.

> ### Support for multiple architectures

Reviewer 3: question, does SDK support multi-arch operator builds currently in any way?

Author: Not really. There is nothing in the SDK preventing the release of operators running on multiple architectures but to the best of my knowledge there is nothing supporting it either. With "Support" I meant here the operator not the SDK. I could rename the chapter to "Multiple architectures" or "Operator support for multiple architecture" as it seems confusing.

Author: Thinking of it. I could also reference the docker way of doing multi-arch builds. If we wanted the SDK to support multi-arch builds of the operator the Makefile would need to be amended for it. Also the generated Dockerfile for go operators is explicitly referencing GOARCH=amd64. With docker TARGETPLATFORM could be used, which is  automatically set to what is provided by the --platform flag of buildx. This does not come out of the box with buildah yet.

Reviewer 4: Agreed that Makefile and dockerfile could have amends such as setting `--platform` flag to `TARGETPLATFORM`. 

And modify the `Makefile` to build the docker image, the command will be: 
`docker buildx build --platform linux/amd64,linux/arm64 -t ${IMG}`.

Also, `--platform` is a docker buildx feature, I am not sure that it work with `docker build`.

Author: Yes multi-arch builds is supported by docker buildx not docker build. I would propose the following:
1. with this PR amending the documentation for the state as is
2. Creating an issue where we can discuss operator-sdk enhancements. I think that operator-sdk init could be extended to generate Makefile and Dockerfile with multi-arch capabilities based on a flag. Another similar thing would be to have the Makefile using podman/buildah rather than docker based on another flag. It seems that the option was previously available with --image-builder but disappeared when the build moved to Makefile (the user needs now to update the Makefile manually)

Author: I created this feature request: https://github.com/operator-framework/operator-sdk/issues/5180

```shell
docker buildx build --push --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x --tag registry/username/repo:v1 .
```

See [docker documentation](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) for additional options.
 
**Caveats**: the Dockerfile generated by the SDK for the operator explicitely references GOARCH=amd64 for go build. This can be amended to GOARCH=$TARGETPLATFORM. Docker will automatically set the environment variable to the value specified by --platform. With buildah --build-arg will need to be used for the purpose.

All image references in the ClusterServiceVersion should be manifest lists containing the pointers to the image manifests for the supported architectures.

#### Architecture specific content

Operand images that require different image content (such as unique runtime packages) per architecture should be handled by the operator itself based on environment variables. That is, the CSV can list environment variables for each supported architecture / OS, for example:

```
MY_OPERAND_ARM64: my/arm64/image
MY_OPERAND_PPC64LE: my/ppc64le/image

> ```

Reviewer 2: Thinking about this more, can't this be accomplished with manifest lists only? The arch-specific images that a manifest list points to can be completely different (in fact, they already are since they have different platform-specific binaries).

```shell
docker buildx build --push --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x --tag registry/username/repo:v1 .
```

See [docker documentation](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) for additional options.
 
Note: the Dockerfile generated by the SDK for the operator explicitely references GOARCH=amd64 for go build. This can be amended to GOARCH=$TARGETPLATFORM. Docker will automatically set the environment variable to the value specified by --platform. With buildah --build-arg will need to be used for the purpose.

All image references in the ClusterServiceVersion should be manifest lists containing the pointers to the image manifests for the supported architectures.

#### Architecture specific content

Operand images that require different image content (such as unique runtime packages) per architecture should be handled by the operator itself based on environment variables. That is, the CSV can list environment variables for each supported architecture / OS, for example:

```
MY_OPERAND_ARM64: my/arm64/image
MY_OPERAND_PPC64LE: my/ppc64le/image
```

> Based on the architecture and OS of the container the operator is running in, the operator can decide which environment variable to use to deploy the correct operands. When using a go-based operator the GOARCH environment variable in the container can provide the required information.

Reviewer 2: This seems like it would only work it homogeneous clusters. What about clusters that have different nodes with different architectures where the operator may be running on `amd64` but the operand gets scheduled on `arm64`?

Author: It is a bit of a corner case. Multi-arch clusters are for instance not supported with OpenShift today.
For clusters that have this setup cluster-admins would need either:
- to use taints and toleration
- or to use NodeSelectors leveraging labels set by the Node Feature Discovery for instance
so that operators and operands run on the same architecture. This is however not driven by the operator authors.

The only (and preferred) approach that really works is with manifest lists as the container engine would pull the right image depending on the architecture.

I can add a caveat to highlight the scenario you are referring to.

Author: As discussed I removed the paragraph

#### Bundle image

> The bundle image only gets built for x86_64 and, since it does not contain any RPM content, it will be able to run on any architecture.

Reviewer 2: This sentence is a bit confusing. It mentions being built for x86_64, references RPMs(?), and then says it can run on any architecture. I think we could simplify this to add more clarity:

```suggestion
Bundle images are not architecture-specific. They contain only plaintext kubernetes manifests and operator metadata.
```

Author: Thanks. I will use your wording.

#### ClusterServiceVersion labels

> Since all indexes for all architectures share the same content, installable operators are filtered in the UI based on the arch labels in the CSV of the head of the default channel.

Reviewer 2: What UI?

If this is referencing features specific to the OpenShift Console, I don't think it belongs in upstream documentation.

Author: Ack. I can rephrase it and take over the couple of lines from OLM documentation:
https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/#caveats

Labels for OS and architectures can be set in the CSV with:

```
labels:
    operatorframework.io/arch.<GOARCH>: supported

>     operatorframework.io/os.<GOOS>: supported

Reviewer 2: How does this work?

If I have an operator that supports "windows/amd64" and "linux/arm64" only, what labels would I set? It seems like I would set:

```yaml
labels:
    operatorframework.io/arch.amd64: supported
    operatorframework.io/os.windows: supported
    operatorframework.io/arch.arm64: supported
    operatorframework.io/os.linux: supported
```

But that makes it look like all combinations of arch/os are supported (i.e. `windows/amd64`, `windows/arm64`, `linux/amd64`, and `linux/arm64`).

Author: You are inferring a coupling between architecture and OS that may not have been foreseen in OLM. To be honest I am not sure that someone would do that: supporting one architecture with an OS and another architecture with another OS but feel free to open an OLM issue: https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/#multiple-architectures

Reviewer 2: Yeah it seems like we really need OLM to deprecate these labels and replace them with something like this:
```yaml
labels:
    operatorframework.io/platform.linux/amd64: supported
    operatorframework.io/platform.windows/arm64: supported
```

And if we need to do that, I'm not sure it makes sense to document multi-arch support in the SDK docs until this is resolved.

Reviewer 2: /hold

Author: There is a need today for providing documentation on how things are working right now. When/if labeling gets improved the documentation can reflect the new state of the world. This is just business as usual: code/api changes need to come together with the matching documentation updates.
I don't see why documentation that is already written (I mean this PR) needs to be set on hold for something that does not exist, needs to be implemented with the old version needing to go through a deprecation process.

Author: For the sake of progressing this PR and getting instructions for multi-arch builds in the documentation I have removed the label part of the PR and replaced it with a link to the OLM documentation. I believe it is nicer for the reader not to have to jump between pages and to get everything in one place but it is still better than having nothing.
For the records OLM does not do anything today with the architecture and os labels. It is only used by the OpenShift console for filtering.
